File: 278. First Bad Version.java
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        if (n <= 0) {
            return -1;
        }
        
        int left = 1;
        int right = n;
        
        if (n == 1) {
            return 1;
        }
        
        while(left < right) {
            int mid = left + (right - left) / 2;
            
            if (isBadVersion(mid)) {
                if (!isBadVersion(mid-1)) {
                    return mid;
                }
                right = mid;
                
            } else {
                left = mid + 1;
            }
        }
        
        return left;
    }
}

File: 278. First Bad Version.py
class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """

        start = 0
        end = n
        
        while start + 1 < end:
            mid = int(start + (end - start) / 2)
            
            if isBadVersion(mid):
                end = mid
            else:
                start = mid
        
        if isBadVersion(start):
            return start
        
        if isBadVersion(end):
            return end
        
        return 0
